api/valor_api/backend/metrics/text_generation.py

```diff
@@ -38,6 +41,147 @@
+        references=processed_references,
+        rouge_types=rouge_types,
+        use_stemmer=use_stemmer,
+        use_aggregator=False,  # aggregation gives us an average across all predicitons, which isn't what we want
+    )
+
+    assert metrics is not None  # handle type error
+
+    # find the max value for each prediction
+    output = defaultdict(lambda: defaultdict(float))
+    for i, prediction in enumerate(processed_prediction):
+        for type_ in rouge_types:
+            output[prediction][type_] = max(
+                metrics[type_][i], output[prediction][type_]
+            )
+
+    return [
+        {"prediction": prediction, "value": dict(value)}
+        for prediction, value in output.items()
+    ]
+
+
+def _calculate_sentence_bleu(
+    predictions: str | list[str],
+    references: list[str] | list[list[str]],
+    weights: list[float] = [0.25, 0.25, 0.25, 0.25],
+) -> list[dict]:
+    """
+    Calculate sentence BLEU scores for a set of prediction-groundtruth pairs.
+
+    Parameters
+    ----------
+    predictions: str | list[str]
+        The predictions to score. Each prediction should be a string with tokens separated by spaces.
+    references: list[str] | list[list[str]
+        A list of reference for each prediction or a list of several references per prediction. Each reference should be a string with tokens separated by spaces.
+    weights: list[float]
+        The default BLEU calculates a score for up to 4-grams using uniform
+        weights (this is called BLEU-4). To evaluate your translations with
+        higher/lower order ngrams, use customized weights. Example: when accounting
+        for up to 5-grams with uniform weights (this is called BLEU-5) use [1/5]*5
+    """
+    if (
+        not predictions
+        or not references
+        or not weights
+        or isinstance(references, str)
+        or len(weights) == 0
+    ):
+        raise ValueError(
+            "Received incorrect inputs. predictions should be a string, references a list of strings, and weights a list/tuple of floats"
+        )
+
+    # handle case where user passes in a single prediction
+    if isinstance(predictions, str):
+        processed_predictions = [predictions]
+        processed_references = [references]
+        # handle case where user passes multiple predictions
+    elif isinstance(predictions, list) and all(
+        [isinstance(lst, list) for lst in references]
+    ):
+        processed_predictions = predictions
+        processed_references = references
+    else:
+        raise ValueError(
+            "prediction should be a str or list[str]. If prediction is a list[str], then references must be a list of lists."
+        )
+
+    # find the max value for each prediction
+    output = defaultdict(float)
+    for pred, refs in zip(processed_predictions, processed_references):
+
+        split_prediction = pred.split()
+        split_references = [ref.split() for ref in refs]  # type: ignore
```

Comment:
In the data drift research project, we split text using a tokenizer, instead of just splitting by word. It seems to me that using a tokenizer would be better in most use cases, as a tokenization of a piece of text is a more meaningful representation than splitting by word. However on the data drift project, Nanbo and I had more freedom to explore and implement as we thought seemed best as we were experimenting. If splitting by word is the more common or default approach, then we could leave this as is.

Reply:
thanks for this comment! I generally agree that using a tokenizer makes more sense here. I used `split` originally since that was used in the hugging face example, but I can use `nltk.tokenize.word_tokenize` instead without adding another dependency. will make that change before merging